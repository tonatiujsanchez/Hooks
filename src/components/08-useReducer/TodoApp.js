
import { useEffect, useReducer } from 'react'
import TodoHeader from './TodoHeader'
import TodoList from './TodoList'
import todoReducer from './todoReducer'

import './todos.css'



const init = () => {

    if (localStorage.getItem('todos-react-reducer')) {
        return JSON.parse(localStorage.getItem('todos-react-reducer'))
    } else {
        return []
    }
}


const TodoApp = () => {


    const [todos, dispatch] = useReducer(todoReducer, [], init)

    useEffect(() => {
        localStorage.setItem('todos-react-reducer', JSON.stringify(todos))
    }, [todos])


    const handleDelete = (todoId) => {
        dispatch({
            type: 'delete',
            paylod: todoId
        })
    }

    const handleToggle = ( todoId ) =>{
        dispatch({
            type: 'toggle',
            paylod: todoId
        })
    }

    const handleAddTodo = ( newTodo ) =>{
        dispatch({
            type: 'add',
            paylod: newTodo
        })
    }

    return (
        <div className="row">
            <div className="col-md-10 col-lg-8 mx-auto">
                <TodoHeader todos={ todos } handleAddTodo={ handleAddTodo } />
                <hr />

                <TodoList 
                    todos={ todos } 
                    handleToggle={ handleToggle } 
                    handleDelete={ handleDelete } />
            </div>
        </div>

    )
}

export default TodoApp