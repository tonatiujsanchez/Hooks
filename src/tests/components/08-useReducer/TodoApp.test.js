import { act } from "@testing-library/react"
import { mount, shallow } from "enzyme"
import TodoApp from "../../../components/08-useReducer/TodoApp"
import { demoTodos } from "../../fixtures/demoTodos"

describe('Pruebas en <TodoApp />', () => { 

    const wrapper = shallow( <TodoApp /> )
    Storage.prototype.setItem = jest.fn(()=>{})

    test('Debe de mostrarse correctamente', () => { 
        expect( wrapper ).toMatchSnapshot()
    })


    test('Debe de agregar el Todo', ()=>{

        const newTodo = {
            id: expect.any(Number),
            desc: 'Aprender Firebase',
            done: false
        }

        const wrapper = mount( <TodoApp /> )
        const handleAddTodo = wrapper.find('TodoHeader').prop('handleAddTodo')
        
        act(()=> {
             handleAddTodo(newTodo)
             handleAddTodo(demoTodos[0])
        })

        const small = wrapper.find('h1 small')

        expect( small.text().trim() ).toBe(`( 2 )`)

        expect( localStorage.setItem ).toHaveBeenCalledTimes(2)

    })

    
   test('Debe de eliminar el Todo', () => { 
        const wrapper = mount( <TodoApp /> )
        const handleAddTodo = wrapper.find('TodoHeader').prop('handleAddTodo')
        const handleDelete = wrapper.find('TodoList').prop('handleDelete')
        act(()=> {
            handleAddTodo(demoTodos[0])
            handleDelete(demoTodos[0].id)
       })

       const small = wrapper.find('h1 small')

       expect( small.text().trim() ).toBe(`( 0 )`)
    })


 })