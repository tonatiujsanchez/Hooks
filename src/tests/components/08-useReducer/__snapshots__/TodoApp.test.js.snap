// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe de mostrarse correctamente 1`] = `
<div
  className="row"
>
  <div
    className="col-md-10 col-lg-8 mx-auto"
  >
    <TodoHeader
      handleAddTodo={[Function]}
      todos={Array []}
    />
    <hr />
    <TodoList
      handleDelete={[Function]}
      handleToggle={[Function]}
      todos={Array []}
    />
  </div>
</div>
`;
